### ISSUE TITLE
RuntimeError for batch size greater than 1 in DistriPixArtAlphaPipeline - Reopen

### ISSUE THREAD

**Issue by SaudxInu** [xdit-project/xDiT#98]
#81 

Hi,

Error still persist and it only works after commenting out the flagged lines of code.

After commenting out multi GPUs and batch size works for all parallelism except pipefusion.

For the following code line to run the script,

`torchrun --nproc_per_node=2 scripts/pixart_example.py -p pipefusion`

I get the following error,

```
[rank1]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/pipelines/pip/distri_pixartalpha.py", line 242, in __call__
[rank1]: AttributeError: 'NoneType' object has no attribute 'classify_height_width_bin'
```

If i comment out the line 242 in `PipeFusion/pipefuser/pipelines/pip/distri_pixartalpha.py`, it works.

Also, there is some dependency conflict between "diffusers==0.29.0" and "distvae==0.0.0b3" (requires "diffusers==0.27.2").

@Steaunk @feifeibear

**Comment by Steaunk** [comment]
Hi, SaudxInu. We are working on the solving dependency conflict now.

**Comment by feifeibear** [comment]
> #81
> 
> Hi,
> 
> Error still persist and it only works after commenting out the flagged lines of code.
> 
> After commenting out multi GPUs and batch size works for all parallelism except pipefusion.
> 
> For the following code line to run the script,
> 
> `torchrun --nproc_per_node=2 scripts/pixart_example.py -p pipefusion`
> 
> I get the following error,
> 
> ```
> [rank1]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/pipelines/pip/distri_pixartalpha.py", line 242, in __call__
> [rank1]: AttributeError: 'NoneType' object has no attribute 'classify_height_width_bin'
> ```
> 
> If i comment out the line 242 in `PipeFusion/pipefuser/pipelines/pip/distri_pixartalpha.py`, it works.
> 
> Also, there is some dependency conflict between "diffusers==0.29.0" and "distvae==0.0.0b3" (requires "diffusers==0.27.2").
> 
> @Steaunk @feifeibear

Sorry, could you please pull the lastest main branch. We recently change code frequently.

**Comment by SaudxInu** [comment]
Sure. Give me 30mins to check the fix.

**Comment by SaudxInu** [comment]
Hi,

I pulled the latest code from main. 

1. Dependency conflict is fixed.
2. `self.image_processor.classify_height_width_bin` not found is fixed.
2. All parallelism configs are running fine with different batch size and number of GPUs only after commenting pre-run lines flagged in the original issue.

**Comment by SaudxInu** [comment]
Adding again for reference.

If I comment out the following lines (282, 283, 288, 289) in PipeFusion/pipefuser/pipelines/pixartalpha.py the script is running fine for number of GPUs 1 and all parallelism except pipefusion.

```python
pipeline.transformer.set_counter(0)
pipeline.transformer(**static_inputs, return_dict=False, record=True)
```

I think there is some bug in static inputs.

**Comment by Steaunk** [comment]
Hi, @SaudxInu  I cannot reproduce your results in parallelism=pipefusion. Could you please show me some more details?

Here is the script I use: https://github.com/PipeFusion/PipeFusion/blob/batch-size%3D2/scripts/pixart_example.py
And cmd: `torchrun --nproc_per_node=2 scripts/pixart_example.py -p pipefusion --pp_num_patch 4 --output_type pil`

**Comment by SaudxInu** [comment]
In https://github.com/PipeFusion/PipeFusion/blob/batch-size%3D2/scripts/pixart_example.py please make the following changes,

```python
parser.add_argument(
    "--prompt",
    type=str,
    default=["An astronaut riding a green horse", "An astronaut riding a green horse"],
)
...
distri_config = DistriConfig(
        height=args.height,
        width=args.width,
        warmup_steps=args.pipefusion_warmup_step,
        do_classifier_free_guidance=True,
        split_batch=False,
        parallelism=args.parallelism,
        mode=args.sync_mode,
        pp_num_patch=args.pp_num_patch,
        use_resolution_binning=not args.no_use_resolution_binning,
        use_cuda_graph=args.use_cuda_graph,
        attn_num=args.attn_num,
        scheduler=args.scheduler,
        ulysses_degree=args.ulysses_degree,
        batch_size=2,
 )
```

Then run,

```shell
torchrun scripts/pixart_example.py
```

You will get the following error,

```shell
[rank0]: RuntimeError: The size of tensor a (2) must match the size of tensor b (4) at non-singleton dimension 0
```

Now if you comment out the following lines (282, 283, 288, 289) in PipeFusion/pipefuser/pipelines/pixartalpha.py the script is running fine for all batch sizes, num GPUs and parallel configs.

```python
# Used to create communication buffer
comm_manager = None
if distri_config.n_device_per_batch > 1:
    comm_manager = PatchParallelismCommManager(distri_config)
    pipeline.transformer.set_comm_manager(comm_manager)

    # Only used for creating the communication buffer
    # pipeline.transformer.set_counter(0)
    # pipeline.transformer(**static_inputs, return_dict=False, record=True)
    if comm_manager.numel > 0:
        comm_manager.create_buffer()

# Pre-run
# pipeline.transformer.set_counter(0)
# pipeline.transformer(**static_inputs, return_dict=False, record=True)
```

I think there is some bug in static inputs.

**Comment by Steaunk** [comment]
Yeah. As you can see, I've already added `batch_size=2` & `prompt=[...,...]` in https://github.com/PipeFusion/PipeFusion/blob/batch-size%3D2/scripts/pixart_example.py

And I can't find this bug. Can you try this script?

**Comment by SaudxInu** [comment]
I pulled from this branch.

Same error.

As soon as I comment the flagged lines, everything works great. I even checked the images, they are correct.

Maybe try installing the repo from scratch. Like I did that mulitple times and get the same error.

**Comment by SaudxInu** [comment]
Pasting complete trace.

```shell
(pipe_fusion) saudi@pipe-fusion-debug-0-0:~$ torchrun scripts/pixart_example.py --output_type pil
/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/diffusers-0.29.0-py3.9.egg/diffusers/models/transformers/transformer_2d.py:34: FutureWarning: `Transformer2DModelOutput` is deprecated and will be removed in version 1.0.0. Importing `Transformer2DModelOutput` from `diffusers.models.transformer_2d` is deprecated and this will be removed in a future version. Please use `from diffusers.models.modeling_outputs import Transformer2DModelOutput`, instead.
  deprecate("Transformer2DModelOutput", "1.0.0", deprecation_message)
yunchang not found
WARNING 06-26 10:53:54 [attn.py:30] ring flash attn not found
WARNING 06-26 10:53:54 [attn.py:32] ring flash attn not found
Some weights of the model checkpoint were not used when initializing PixArtTransformer2DModel: 
 ['caption_projection.y_embedding']
INFO 06-26 10:53:55 [distri_dit_pp.py:63] Not using parallelism for DiT
DistriPixArtAlphaPipeline from pretrain stage 1 0.0 GB
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  2.41it/s]
Loading pipeline components...:  40%|████████████████████████████████████████▍                                                            | 2/5 [00:00<00:01,  2.19it/s]Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Loading pipeline components...: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  4.46it/s]
DistriPixArtAlphaPipeline from pretrain stage 2 12.947305984 GB
[rank0]: Traceback (most recent call last):
[rank0]:   File "/sensei-fs/users/saudi/internships_24/gpu_opt/PipeFusion/scripts/pixart_example.py", line 173, in <module>
[rank0]:     main()
[rank0]:   File "/sensei-fs/users/saudi/internships_24/gpu_opt/PipeFusion/scripts/pixart_example.py", line 141, in main
[rank0]:     pipeline = DistriPixArtAlphaPipeline.from_pretrained(
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/pipelines/pixartalpha.py", line 126, in from_pretrained
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/pipelines/pixartalpha.py", line 57, in __init__
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/utils/_contextlib.py", line 115, in decorate_context
[rank0]:     return func(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/pipelines/pixartalpha.py", line 289, in prepare
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/models/distri_dit_pp.py", line 132, in forward
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/pipefusion-0.1-py3.9.egg/pipefuser/modules/dit/patch_parallel/transformer_2d.py", line 136, in forward
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/diffusers-0.29.0-py3.9.egg/diffusers/models/normalization.py", line 116, in forward
[rank0]:     embedded_timestep = self.emb(timestep, **added_cond_kwargs, batch_size=batch_size, hidden_dtype=hidden_dtype)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
[rank0]:     return self._call_impl(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/nn/modules/module.py", line 1541, in _call_impl
[rank0]:     return forward_call(*args, **kwargs)
[rank0]:   File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/diffusers-0.29.0-py3.9.egg/diffusers/models/embeddings.py", line 1033, in forward
[rank0]:     conditioning = timesteps_emb + torch.cat([resolution_emb, aspect_ratio_emb], dim=1)
[rank0]: RuntimeError: The size of tensor a (2) must match the size of tensor b (4) at non-singleton dimension 0
E0626 10:54:06.276428 139753597114176 torch/distributed/elastic/multiprocessing/api.py:826] failed (exitcode: 1) local_rank: 0 (pid: 5174) of binary: /opt/conda/envs/pipe_fusion/bin/python
Traceback (most recent call last):
  File "/opt/conda/envs/pipe_fusion/bin/torchrun", line 33, in <module>
    sys.exit(load_entry_point('torch==2.3.1', 'console_scripts', 'torchrun')())
  File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/distributed/elastic/multiprocessing/errors/__init__.py", line 347, in wrapper
    return f(*args, **kwargs)
  File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/distributed/run.py", line 879, in main
    run(args)
  File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/distributed/run.py", line 870, in run
    elastic_launch(
  File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/distributed/launcher/api.py", line 132, in __call__
    return launch_agent(self._config, self._entrypoint, list(args))
  File "/opt/conda/envs/pipe_fusion/lib/python3.9/site-packages/torch-2.3.1-py3.9-linux-x86_64.egg/torch/distributed/launcher/api.py", line 263, in launch_agent
    raise ChildFailedError(
torch.distributed.elastic.multiprocessing.errors.ChildFailedError: 
============================================================
scripts/pixart_example.py FAILED
------------------------------------------------------------
Failures:
  <NO_OTHER_FAILURES>
------------------------------------------------------------
Root Cause (first observed failure):
[0]:
  time      : 2024-06-26_10:54:06
  host      : pipe-fusion-debug-0-0
  rank      : 0 (local_rank: 0)
  exitcode  : 1 (pid: 5174)
  error_file: <N/A>
  traceback : To enable traceback see: https://pytorch.org/docs/stable/elastic/errors.html
============================================================
```

**Comment by feifeibear** [comment]
@SaudxInu Hello, we recently refactored the repo.

could you try the new pixart examples?

**Comment by SaudxInu** [comment]
@feifeibear Sure.

**Comment by SaudxInu** [comment]
@feifeibear Config for Distrifusion?

**Comment by Eigensystem** [comment]
We didn't implement distrifusion in the refactor version.

**Comment by Eigensystem** [comment]
To use dp, you can simply use the scripts in `examples` folder and set multiple prompts in --prompt option, and set --dp_degree option to the number you want.
ps. Please make sure that the number of prompts is greater than dp_degree to make the data parallel available

**Comment by SaudxInu** [comment]
Thanks. I am testing it right now. i will let you know if I encounter any errors.

**Comment by Eigensystem** [comment]
> We didn't implement distrifusion in the refactor version.

The reasons why we didn't implement distrifusion in the new version of hybrid parallelism are as follows:

Distrifusion is not the most efficient parallel method in various situations. For NVLink devices, sp (spatial parallelism) is more memory-efficient and faster. For PCIe devices or cross-node communication, the PipeFusion parallel method is the most efficient. You can refer to the data in the README.md file for more information.

Introducing distrifusion into the current version of hybrid parallelism would significantly increase the complexity of model and latents processing, bringing unnecessary overhead for future model additions.

**Comment by SaudxInu** [comment]
Cool. I just wanted to know for benchmarking and future work explorations.

**Comment by SaudxInu** [comment]
`OSError: Cannot load model models/PixArt-XL-2-1024-MS: model is not cached locally and an error occurred while trying to fetch metadata from the Hub. Please check out the root cause in the stacktrace above.`

Pretrained model path is wrong.

**Comment by SaudxInu** [comment]
Model path is wrong.

`OSError: Cannot load model models/PixArt-XL-2-1024-MS: model is not cached locally and an error occurred while trying to fetch metadata from the Hub. Please check out the root cause in the stacktrace above.`

Even after fixing the error, I am getting the following error.

`ValueError: Non-consecutive added token '<pad>' found. Should have index 32100 but has index 0 in saved vocabulary.`

**Comment by feifeibear** [comment]
> Model path is wrong.
> 
> `OSError: Cannot load model models/PixArt-XL-2-1024-MS: model is not cached locally and an error occurred while trying to fetch metadata from the Hub. Please check out the root cause in the stacktrace above.`
> 
> Even after fixing the error, I am getting the following error.
> 
> `ValueError: Non-consecutive added token '<pad>' found. Should have index 32100 but has index 0 in saved vocabulary.`

You should check if you download the model to disk correctly. Perhaps need to check the hash for the files.

**Comment by SaudxInu** [comment]
https://github.com/xdit-project/xDiT/blob/e7e10001ad769f6128fcf5af706063f0ae6ebc5a/xfuser/model_executor/pipelines/pipeline_pixart_alpha.py#L38

So, I printed the model path. It is 'models/PixArt-XL-2-1024-MS'. I think on your setup you in huggingface cache folder you have named the model like that. So, if it is not present it will try to download it from huggingface. But on huggingface model name is 'PixArt-alpha/PixArt-XL-2-1024-MS'.

In legacy code, path was correct.

**Comment by feifeibear** [comment]
> https://github.com/xdit-project/xDiT/blob/e7e10001ad769f6128fcf5af706063f0ae6ebc5a/xfuser/model_executor/pipelines/pipeline_pixart_alpha.py#L38
> 
> So, I printed the model path. It is 'models/PixArt-XL-2-1024-MS'. I think on your setup you in huggingface cache folder you have named the model like that. So, if it is not present it will try to download it from huggingface. But on huggingface model name is 'PixArt-alpha/PixArt-XL-2-1024-MS'.
> 
> In legacy code, path was correct.

After you correct the huggingface model name, could you run xDiT successfully?

**Comment by SaudxInu** [comment]
I am running this command.

`torchrun --nproc_per_node=4 examples/pixartalpha_example.py --model models/PixArt-XL-2-1024-MS --pipefusion_parallel_degree 2 --num_inference_steps 20 --warmup_steps 0 --prompt "A small dog" --use_split_batch`

It is working fine.

<img width="1013" alt="Screenshot 2024-07-30 at 6 33 18 PM" src="https://github.com/user-attachments/assets/2f174874-c821-458b-83f3-39ff6375fbcd">
